### PR TITLE
mds: add option mds_bal_overload_epochs

### PR DESCRIPTION
**Pull request by zhsgao**
Add an option to configure the number of epochs the overload lasts before migrating,
setting it to a higher value can avoid frequent migrations caused by load fluctuations.

Signed-off-by: Zhansong Gao <zhsgao@hotmail.com>

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "pacific"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

## Checklist
- Tracker (select at least one)
  - [ ] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [ ] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [ ] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [ ] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
</details>

### CODE REVIEW

**Comment by vshankar** [comment]
> Add an option to configure the number of epochs the overload lasts before migrating.

Could you elaborate about the intended use this config option @zhsgao?

**Comment by zhsgao** [comment]
> > Add an option to configure the number of epochs the overload lasts before migrating.
> 
> Could you elaborate about the intended use this config option @zhsgao?

Use this option to adjust the duration of mds overload before migration, setting it to a higher value can reduce frequent migrations caused by load fluctuations.

File: src/common/options/mds.yaml.in [hunk 1]
```diff
@@ -819,6 +819,15 @@ options:
...
   - mds
   with_legacy: true
+# must be overloaded for more than these epochs before we export anything
+- name: mds_bal_overload_epochs
+  type: int
+  level: dev
+  default: 2
+  services:
+  - mds
+  fmt_desc: The number of epochs the overload lasts before Ceph migrates.
```
Comment (vshankar) [inline src/common/options/mds.yaml.in:829]: Please add a long description for the config.

**Review by vshankar — COMMENTED** [review]
@zhsgao I'm curious is you ran into an issue that prompted this hardcoded value to be made configurable?

**Comment by zhsgao** [comment]
> @zhsgao I'm curious is you ran into an issue that prompted this hardcoded value to be made configurable?

The loads of multiple MDS daemons are unbalanced and fluctuate greatly in a short period of time in my Ceph cluster, and migration is very frequent when using the default configurations. So I set `mds_bal_min_rebalance` to a larger value to reduce migration(large enough can disable it), it decides whether to migrate based on the degree of load imbalance. So I think there should be an option to decide whether to migrate based on the duration of load imbalance.

**Comment by vshankar** [comment]
> > @zhsgao I'm curious is you ran into an issue that prompted this hardcoded value to be made configurable?
> 
> The loads of multiple MDS daemons are unbalanced and fluctuate greatly in a short period of time in my Ceph cluster, and migration is very frequent when using the default configurations. So I set `mds_bal_min_rebalance` to a larger value to reduce migration(large enough can disable it), it decides whether to migrate based on the degree of overload. So I think there should be an option to decide whether to migrate based on the duration of overload.

Fair enough. Please add a long desc to the option. Otherwise LGTM.

**Comment by zhsgao** [comment]
> > > @zhsgao I'm curious is you ran into an issue that prompted this hardcoded value to be made configurable?
> > 
> > 
> > The loads of multiple MDS daemons are unbalanced and fluctuate greatly in a short period of time in my Ceph cluster, and migration is very frequent when using the default configurations. So I set `mds_bal_min_rebalance` to a larger value to reduce migration(large enough can disable it), it decides whether to migrate based on the degree of overload. So I think there should be an option to decide whether to migrate based on the duration of overload.
> 
> Fair enough. Please add a long desc to the option. Otherwise LGTM.

I have modified `fmt_desc` and the commit message.

**Comment by vshankar** [comment]
jenkins retest this please

**Review by vshankar — APPROVED** [review]
https://tracker.ceph.com/projects/cephfs/wiki/Main#26-Sep-2023

**Pull request merged**